File: examples/data_processing.exs
# Data processing example
# Run: TYREX_BUILD=true mix run examples/data_processing.exs

IO.puts("=== Tyrex Data Processing Examples ===\n")

{:ok, pid} = Tyrex.start(main_module_path: "examples/data_processing.js")

# CSV parsing
csv = "name,age,city\nAlice,30,NYC\nBob,25,LA\nCharlie,35,Chicago"
{:ok, result} = Tyrex.eval("parseCSV(`#{csv}`)", pid: pid)
IO.puts("CSV parsed:")
for row <- result, do: IO.puts("  #{row["name"]} (#{row["age"]}) from #{row["city"]}")

# Statistics
{:ok, result} = Tyrex.eval("stats([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])", pid: pid)
IO.puts("\nStatistics for 1..10:")

IO.puts(
  "  Mean: #{result["mean"]}, Median: #{result["median"]}, StdDev: #{Float.round(result["stddev"], 2)}"
)

# URL parsing
{:ok, result} =
  Tyrex.eval(~s|parseUrl("https://example.com/api/users?page=2&limit=10")|, pid: pid)

IO.puts("\nURL parsed:")
IO.puts("  Host: #{result["host"]}, Path: #{result["pathname"]}")
IO.puts("  Params: #{inspect(result["params"])}")

# HTML sanitization
{:ok, result} = Tyrex.eval(~s|sanitizeHtml('<script>alert("xss")</script>')|, pid: pid)
IO.puts("\nSanitized: #{result}")

# Data transformation
data =
  Jason.encode!([
    %{first_name: "Alice", last_name: "Smith"},
    %{first_name: "Bob", last_name: "Jones"}
  ])

mapping = Jason.encode!(%{name: "first_name", surname: "last_name"})
{:ok, result} = Tyrex.eval("transformData(#{data}, #{mapping})", pid: pid)
IO.puts("\nTransformed data:")
for row <- result, do: IO.puts("  #{row["name"]} #{row["surname"]}")

Tyrex.stop(pid: pid)
IO.puts("\n=== Done! ===")

File: examples/data_processing.js
// Example: Data processing utilities loaded as main module
// Demonstrates using JS libraries for data transformation

// CSV parser (simple)
globalThis.parseCSV = (csv) => {
  const lines = csv.trim().split("\n");
  const headers = lines[0].split(",").map((h) => h.trim());
  return lines.slice(1).map((line) => {
    const values = line.split(",").map((v) => v.trim());
    return Object.fromEntries(headers.map((h, i) => [h, values[i]]));
  });
};

// JSON transformer
globalThis.transformData = (data, mapping) => {
  if (typeof data === "string") data = JSON.parse(data);
  if (typeof mapping === "string") mapping = JSON.parse(mapping);

  return data.map((item) => {
    const result = {};
    for (const [newKey, oldKey] of Object.entries(mapping)) {
      result[newKey] = item[oldKey];
    }
    return result;
  });
};

// Statistics
globalThis.stats = (numbers) => {
  if (typeof numbers === "string") numbers = JSON.parse(numbers);
  const sorted = [...numbers].sort((a, b) => a - b);
  const sum = sorted.reduce((a, b) => a + b, 0);
  const mean = sum / sorted.length;
  const median =
    sorted.length % 2 === 0
      ? (sorted[sorted.length / 2 - 1] + sorted[sorted.length / 2]) / 2
      : sorted[Math.floor(sorted.length / 2)];
  const variance =
    sorted.reduce((acc, val) => acc + Math.pow(val - mean, 2), 0) /
    sorted.length;
  return {
    count: sorted.length,
    sum,
    mean,
    median,
    min: sorted[0],
    max: sorted[sorted.length - 1],
    stddev: Math.sqrt(variance),
  };
};

// String sanitization
globalThis.sanitizeHtml = (html) => {
  return html
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#039;");
};

// URL parsing
globalThis.parseUrl = (url) => {
  const parsed = new URL(url);
  return {
    protocol: parsed.protocol,
    host: parsed.host,
    pathname: parsed.pathname,
    search: parsed.search,
    params: Object.fromEntries(parsed.searchParams),
  };
};
